tier based ID rework

TheChosenEvilOne opened the pull request:
## About The Pull Request
This changes all accesses to have a tier from 1 to 5 which tells what is the required tier of ID card to add that access on, for example you can't give an assistant (tier 1 ID card) ID console access (tier 5 access), but you can give maintenance access (tier 1) to a scientist (tier 2 ID card). Due to there being jobs that require special access ID cards also have additional access chips (AA-chips, I only named them that because of the AA), each job that requires access with higher tier than the ID card they have will on roundstart spawn with an AA-chip that contains the access they require. AA-chips can also be removed with a screwdriver and inserted to any ID card that has chip slots by hand (this means you can steal ID console access from the HoP with a screwdriver, because the HoP's ID card is tier 4 and ID console access is tier 5).

The card tiers go as follows:
Tier 0 - Prisoner
Tier 1 - Assistant
Tier 2 - All jobs that are not mentioned
Tier 3 - QM, Warden, Virologist
Tier 4 - CE, CMO, HoP, HoS, RD
Tier 5 - Captain

## Why It's Good For The Game
The current ID trim thing is bad and limiting for no reason.

## Changelog
:cl:
add: ID card tiers, you can only add access to ID cards if the access' tier is lower or equal to the ID card's tier.
add: additional access chips, every job that has a lower tier ID card than some accesses they require spawn with an additional access chip in the ID card which contain the access required. You can use a screwdriver to remove the chip and add a chip in an ID card by clicking an ID card with it.
del: Removed ID trims.
balance: Agent ID card no longer has a limit on how many access levels it can hold.
/:cl:

Canohaaaaaaa commented on _maps/map_files/Deltastation/DeltaStation2.dmm:
```diff
@@ -45875,7 +45875,6 @@
 /obj/effect/turf_decal/tile/neutral{
 	dir = 8
 	},
-/obj/item/paper/fluff/ids_for_dummies,
```
Do you have a replacement for this ?

Canohaaaaaaa commented on code/game/machinery/computer/communications.dm:
```diff
@@ -132,7 +130,7 @@
 					playsound(src, 'sound/machines/terminal_prompt_deny.ogg', 50, FALSE)
 					return
-				if (!(ACCESS_CAPTAIN in id_card.access))
+				if (!(ACCESS_CAPTAIN in id_card.GetAccess()))
```
`snake_case` the ID procs please

Canohaaaaaaa commented on code/game/machinery/computer/communications.dm:
```diff
@@ -313,7 +311,8 @@
 				authenticated = TRUE
-				authorize_access = SSid_access.get_region_access_list(list(REGION_ALL_STATION))
+				// TODO:
+				// authorize_access = SSid_access.get_region_access_list(list(REGION_ALL_STATION))
```
TODO leftover

Canohaaaaaaa commented on code/game/machinery/porta_turret/portable_turret.dm:
```diff
@@ -547,7 +547,7 @@ DEFINE_BITFIELD(turret_flags, list(
 
 	if(turret_flags & TURRET_FLAG_AUTH_WEAPONS) //check for weapon authorization
-		if(isnull(perp.wear_id) || istype(perp.wear_id.GetID(), /obj/item/card/id/advanced/chameleon))
+		if(isnull(perp.wear_id) || istype(perp.wear_id.GetID(), /obj/item/card/id/chameleon))
```
`snake_case`

Canohaaaaaaa commented on code/game/objects/structures/crates_lockers/closets.dm:
```diff
@@ -534,7 +534,7 @@
 			else
 				req_access = list()
-				req_access += pick(SSid_access.get_region_access_list(list(REGION_ALL_STATION)))
+				//TODO: req_access += pick(SSid_access.get_region_access_list(list(REGION_ALL_STATION)))
```
TODO leftover

Canohaaaaaaa commented on code/controllers/subsystem/id_access.dm:
```diff
@@ -335,143 +171,49 @@ SUBSYSTEM_DEF(id_access)
 	return tally
+
+/datum/controller/subsystem/id_access/proc/apply_card_access(obj/item/card/id/id, card_access, chip = FALSE, force = FALSE)
+	var/datum/card_access/C = card_access_instances[card_access]
```
Descriptive var name

TheChosenEvilOne replied:
watch this

Canohaaaaaaa commented on code/game/machinery/PDApainter.dm:
```diff
@@ -40,334 +30,113 @@
-			user.visible_message(span_notice("[user] is repairing [src]."), \
-							span_notice("You begin repairing [src]..."), \
-							span_hear("You hear welding."))
+			user.visible_message("<span class='notice'>[user] is repairing [src].</span>", \
+							"<span class='notice'>You begin repairing [src]...</span>", \
+							"<span class='hear'>You hear welding.</span>")
```
```suggestion
			user.visible_message(span_notice("[user] is repairing [src]."), \
							span_notice("You begin repairing [src]..."), \
							span_hear("You hear welding."))
```

Canohaaaaaaa commented on code/game/machinery/PDApainter.dm:
```diff
@@ -40,334 +30,113 @@
 				if(!(machine_stat & BROKEN))
 					return
-				to_chat(user, span_notice("You repair [src]."))
+				to_chat(user, "<span class='notice'>You repair [src].</span>")
```
```suggestion
				to_chat(user, span_notice("You repair [src]."))
```

Canohaaaaaaa commented on code/game/machinery/PDApainter.dm:
```diff
@@ -40,334 +30,113 @@
-			to_chat(user, span_warning("There is already an ID card inside!"))
+	else if(istype(O, /obj/item/pda))
+		if(storedpda)
+			to_chat(user, "<span class='warning'>There is already a PDA inside!</span>")
```
```suggestion
			to_chat(user, span_warning("There is already a PDA inside!"))
```

Canohaaaaaaa commented on code/game/machinery/PDApainter.dm:
```diff
@@ -40,334 +30,113 @@
+	if(storedpda)
+		if(machine_stat & BROKEN) //otherwise the PDA is stuck until repaired
+			ejectpda()
+			to_chat(user, "<span class='info'>You manage to eject the loaded PDA.</span>")
```
```suggestion
			to_chat(user, span_info("You manage to eject the loaded PDA."))
```

Canohaaaaaaa commented on code/game/machinery/PDApainter.dm:
```diff
@@ -40,334 +30,113 @@
+			storedpda.desc = initial(pda_path.desc)
+			ejectpda()
+	else
+		to_chat(user, "<span class='warning'>[src] is empty!</span>")
```
```suggestion
		to_chat(user, span_warning("[src] is empty!"))
```

Canohaaaaaaa commented on code/game/machinery/PDApainter.dm:
```diff
@@ -40,334 +30,113 @@
+		storedpda = null
+		update_appearance()
+	else
+		to_chat(usr, "<span class='warning'>[src] is empty!</span>")
```
```suggestion
		to_chat(usr, span_warning("[src] is empty!"))
```

TheChosenEvilOne commented on code/controllers/subsystem/id_access.dm:
```diff
@@ -335,143 +171,49 @@ SUBSYSTEM_DEF(id_access)
 
 	return tally
+
+/datum/controller/subsystem/id_access/proc/apply_card_access(obj/item/card/id/id, card_access, chip = FALSE, force = FALSE)
+	var/datum/card_access/C = card_access_instances[card_access]
+	var/list/chip_access = list()
+	id.assignment = C.assignment
+	for (var/tier in C.access)
+		if (!force && id.access_tier < text2num(tier))
+			if (chip)
+				chip_access.Add(C.access[tier])
+			continue
+		id.access.Add(C.access[tier])
+
+	if (chip_access.len)
+		var/obj/item/card_access_chip/roundstart/AA = new(id, C.assignment)
+		AA.access = chip_access
+		id.apply_access_chip(AA)
```
```suggestion
/datum/controller/subsystem/id_access/proc/apply_card_access(obj/item/card/id/id, card_access_path, chip = FALSE, force = FALSE)
	var/datum/card_access/card_access = card_access_instances[card_access_path]
	var/list/chip_access = list()
	id.assignment = card_access.assignment
	for (var/tier in card_access.access)
		if (!force && id.access_tier < text2num(tier))
			if (chip)
				chip_access.Add(card_access.access[tier])
			continue
		id.access.Add(card_access.access[tier])

	if (chip_access.len)
		var/obj/item/card_access_chip/roundstart/AA = new(id, card_access.assignment)
		AA.access = chip_access
		id.apply_access_chip(AA)
```

kriskog commented:
Some stuff:
Requires a chip for weapons permit. This is something the HOS for instance should be able to grant anyone.
No chip slot icon on IDs until you put one on. Removing the chip leaves the slot icon.
No extra chips on station? Was alone one round and had to grab spare, but spare has no chip. Forced to abandon own ID (including bank account) for access. What if the HOP or Captain wishes to promote someone to a head role? Replace someone after an unrecoverable suicide? Prisoners and assistants should be able to get a job and that jobs access if thats how it plays out. You need to figure something out for this. A limited amount of Tier 4 capable chips somewhere secure at least.
Didn't check, but have heads retained the ability to give access inside their departments?

TheChosenEvilOne commented:
well, I already wrote a response to this but forgot to post it so, here I go again.
> * Requires a chip for weapons permit. This is something the HOS for instance should be able to grant anyone.

Giving HoS a way to grant people weapons permit would be good, but I am not sure how it should work. I could give the HoS a box of weapons permit chips or a way to produce weapons permit chips.

> * No chip slot icon on IDs until you put one on. Removing the chip leaves the slot icon.

Fixed this, IDs didn't update overlays when created.

> * No extra chips on station? Was alone one round and had to grab spare, but spare has no chip. Forced to abandon own ID (including bank account) for access. What if the HOP or Captain wishes to promote someone to a head role? Replace someone after an unrecoverable suicide? Prisoners and assistants should be able to get a job and that jobs access if thats how it plays out. You need to figure something out for this. A limited amount of Tier 4 capable chips somewhere secure at least, and Tier 3 chips should be abundant.

I need to figure out how getting higher tier IDs and chips should work, right now what I have been thinking about is that tier 1 and 2 IDs and tier 3 chips should be printable. I could add a box of tier 4 chips in HoP (captain?) locker and possibly allow buying them. Tier 3 IDs as well could be something you could get from cargo.

> * Didn't check, but have heads retained the ability to give access inside their departments?

I added this back, I removed it without realising.

TheChosenEvilOne commented:
I'd say this is about done, I don't really have anything to add to this.
